packages/wonder-blocks-layout/components/layout.js

```diff
@@ -0,0 +1,277 @@
+    mdOrLarger: boolean,
+    /** Should the contents be displayed at a media size of "small" or "medium"? */
+    mdOrSmaller: boolean,
+|};
+
+// If for some reason we're not able to resolve the current media size we
+// fall back to this state.
+const DEFAULT_SIZE = "large";
+
+/**
+ * Layout is responsible for changing the rendering of contents at differently
+ * sized viewports. Layout can be used in conjunction with MediaLayout to
+ * render based on different breakpoint configurations.
+ */
+export default class Layout extends React.Component<
+    Props,
+    {
+        size?: MediaSize,
+    },
+> {
+    watchHandlers: {
+        [query: string]: any,
+    };
+
+    static WATCHERS: {
+        [query: string]: any,
+    } = {};
+
+    static defaultProps = {
+        small: true,
+        medium: true,
+        mdOrLarger: true,
+        mdOrSmaller: true,
+        large: true,
+    };
+
+    state = {
+        size: undefined,
+    };
+
+    componentWillUnmount() {
+        // If we're on the server then we don't have anything to unmount
+        // as we never mounted anything in the first place!
+        if (this.isServerSide()) {
+            return;
+        }
+
+        // We go through the component and remove all of the listeners
+        // that getCurrentSize attached.
+        for (const query of Object.keys(Layout.WATCHERS)) {
+            const watcher = Layout.WATCHERS[query];
+            if (watcher) {
+                const handler = this.watchHandlers[query];
+                watcher.removeListener(handler);
+                delete this.watchHandlers[query];
+            }
+        }
+    }
+
+    getCurrentSize(spec: MediaSpec) {
+        // If we have a state with the current size in it then we always want
+        // to use that. This will happen if the viewport changes sizes after
+        // we've already initialized.
+        if (this.state.size) {
+            return this.state.size;
+        }
+
+        // watchHandlers should never be undefined when state.size is also
+        // undefined, but just in case we fall back to the default size
+        if (this.watchHandlers) {
+            return DEFAULT_SIZE;
+        }
+
+        // We then go through and set up matchMedia watchers for each breakpoint
+        // (if they haven't been created already) and we add listeners to
+        // watch for when the viewport changes size.
+        this.watchHandlers = {};
+
+        for (const size of VALID_MEDIA_SIZES) {
+            if (!spec[size]) {
+                continue;
+            }
+
+            const {query} = spec[size];
+
+            // Don't watch sizes that don't have an associated query
+            if (!query) {
+                continue;
+            }
+
+            // Create a new matchMedia watcher if one doesn't exist yet
+            if (!Layout.WATCHERS[query]) {
+                Layout.WATCHERS[query] = window.matchMedia(query);
```

Comment:
I see, doing it in `componentDidMount` is too late because we need to know the current media size on first render.  We could move this to `componentWillMount` and add a `isServerSide()` check there.  I'll do that in a separate PR.